design/ceph/osd-migration.md

```diff
@@ -0,0 +1,106 @@
+
+## GOALS
+- Newly provisioned clusters should be able to use the specific storage backend while creating OSDs.
+- Existing clusters should be able to upgrade the OSDs to use a specific storage backend, with a minimal downtime and without losing their data.
+
+
+## PROPOSAL
+- Add a new field called `spec.Storage.osdStore` in the ceph cluster yaml.
+- This will contain:
+  - `spec.Storage.osdStore.type`: The type of backend store to be used for OSDs. For example, bluestore, bluestore_rdr, etc. The default store type will be `bluestore`
+  - `spec.Storage.osdStore.updateStore`: Whether the user wants to migrate the existing OSDs to use new storage backend. The migration process will involve destroying the exsisting OSDs one by one, wiping the disks, creating a new OSD and waiting for the PGs to be active+clean before migrating the next OSD. Since this operation involves possible downtime, users should be really sure before proceeding with this action. This field can only take the value as `yes-really-update-store`
+  -
+
+- Example:
+```yaml
+storage:
+  osdStore:
+    type: bluestore_rdr
+    updateStore: yes-really-update-store
+```
+
+- New deployments will use this `osdStore.type` while creating the OSDs.
+- If the user wants to change the `osdStore.type` field for an existing cluster, then they need to provide `spec.Storage.osdStore.updateStore:yes-really-update-store` in order to migrate the existing osds to use update osdStore.
+
+- Add a new label `osd-store:<osd store type>` to all the OSD pods.
+- This label will help to identify the current backend store being used for the OSD and enable the operator to decide if the
+  OSD should be migrated incase the user changes the backend store type in the spec (`spec.Storage.osdStore.type`).
+
+
+### NEW CLUSTERS
+- Currently the osd backend store is hardcoded to `bluestore` throughout the operator code.
+- Add a new environment variable called `ROOK_OSD_STORE_TYPE` in the OSD prepare and OSD pods. This will contain the OSD store type. Use this environment variable to prepare and activate the cluster.
+
+#### OSD Prepare
+- `ceph-volume prepare` command should be updated to use backend store provided in the `spec.storage.osdStore`. If no value is provided for `spec.storage.osdStore` then default to `bluestore` backend.
+
+  - **OSD on disks**:
+    - RAW MODE
+      - `ceph-volume lvm prepare --bluestore_rdr ...`
+    - LWM MODE
+      - `ceph-volume raw prepare --bluestore_rdr ...`
+  - **OSD on PVCs**:
+    -
+
+#### OSD Create
+- `OSD_STORE_FLAG` should use the value provided in the `spec.storage.osdStore`. If no values are provided by the user, then `OSD_STORE_FLAG` should default to `bluestore`
+- `ceph-volume` commands should use the `OSD_STORE_FLAG` while activating the OSDs.
+  - **OSD on disks**:
+    - RAW MODE
+    `ceph-volume raw activate --device "$DEVICE" --no-systemd --no-tmpfs`
```

Comment:
This part is still a bit confusing for me and still talking to ceph team about this to get more answers.  According to what have in rook currently, [no need to pass `OSD_STORE_FLAG` for raw mode.](https://github.com/rook/rook/blob/aee57ea9ffee36d680954193146d1d66c3953167/pkg/operator/ceph/cluster/osd/spec.go#L179) as it uses `bluestore` by default. 

I tried passing the flag with raw mode activation step, but it failed. (Didn't check the logs on why it failed)

According to the cep-volume team, 

- the `activate` flow has an auto detection mechanism to check what objectstore was used. 
- From ceph-volume point of view, the only difference between bluestore and bluestore-rdr is that when it does the bluestore --mkfs, you have to pass it the objectstore you want to use.  And is done during the `prepare` command like below: 
``` Running command: /usr/bin/ceph-osd --cluster ceph --osd-objectstore bluestore-rdr --mkfs -i 0 --monmap /var/lib/ceph/osd/ceph-0/activate.monmap --keyfile - --osd-data /var/lib/ceph/osd/ceph-0/ --osd-uuid 23ad4680-5901-4ba5-ba41-a98c7b285192 --setuser ceph --setgroup ceph```

Reply:
So according to the ceph volume team:
``` Raw mode
 Prepare: pass OSD_STORE flag
 Activate: No need to pass any flag and it will internally find out the objectstore type that was added during prepare. 
 
lvm mode:
 Prepare: pass OSD-STORE flag
 Activate:  Passing the OSD-STORE flag is optional.

Reply:
Ok great, no need to pass the flag to activate for either raw or lvm mode